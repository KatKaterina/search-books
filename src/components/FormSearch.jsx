import React, { useContext, useState, useEffect } from 'react';
import {
  Row,
  Container,
  Col,
  Form,
} from 'react-bootstrap';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import {
  searchBooks,
  setTypeSort,
  setTypeFilter,
  setQuery,
  resetCount,
} from '../slices/booksSlice.js';
import AppContext from '../contexts/index.js';
import FormSelect from './FormSelect.jsx';
import SearchField from './SearchField.jsx';

const FormSearch = () => {
  const dispatch = useDispatch();
  const navigate = useNavigate();

  const [queryText, setQueryText] = useState('');
  const [isTyping, setIsTyping] = useState(false);
  const [filter, setFilter] = useState('all');
  const [sort, setSort] = useState('relevance');
  const [error, setError] = useState(null);

  const { loading } = useSelector((state) => state.books);
  const { stepPagination } = useContext(AppContext);

  const startIndex = 0;
  const statusLoading = loading === 'loading';

  useEffect(() => {
    if (isTyping) {
      return;
    }
    if (queryText.trim() !== '') {
      dispatch(resetCount());
      const param = {
        query: queryText,
        typeFilter: filter,
        typeSort: sort,
        stepPagination,
        startIndex,
      };
      dispatch(searchBooks(param));
      navigate('result');
    }
  }, [filter, sort, queryText, isTyping]);

  const handleSubmit = (e) => {
    e.preventDefault();
    setError(null);
    if (queryText.trim() === '' && e.target.value === '') {
      setError('an empty request is not allowed');
      return;
    }
    setIsTyping(false);
  };

  const handleChange = (e) => {
    const val = e.target.value;
    dispatch(setQuery(val));
    setQueryText(val);
    setIsTyping(true);
  };

  const handleChangeFilter = (e) => {
    const val = e.target.value;
    dispatch(setTypeFilter(val));
    setFilter(val);
  };

  const handleChangeSort = (e) => {
    const val = e.target.value;
    dispatch(setTypeSort(val));
    setSort(val);
  };

  const filterValues = ['all', 'art', 'biography', 'computers', 'history', 'medicine', 'poetry'];
  const sortValues = ['relevance', 'newest'];

  return (
    <Container className="backdrop text-white">
      <Row>
        <Col className="mt-5">
          <h1 className="text-center">Search for books</h1>
        </Col>
      </Row>
      <Row>
        <Form onSubmit={handleSubmit}>
          <SearchField placeholder="e.g. Little Prince" value={queryText} handleChange={handleChange} handleClick={handleSubmit} loading={statusLoading} error={error} />
          <Form.Group as={Row} className="mb-5 mt-5 justify-content-center">
            <FormSelect value={filter} values={filterValues} handleChange={handleChangeFilter} label="Categories" loading={statusLoading} ariaLabel="select-categories" />
            <FormSelect value={sort} values={sortValues} handleChange={handleChangeSort} label="Sorting by" loading={statusLoading} ariaLabel="select-sorting" />
          </Form.Group>
        </Form>
      </Row>
    </Container>
  );
};

export default FormSearch;
