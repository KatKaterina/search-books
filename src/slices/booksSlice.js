import { createSlice, createEntityAdapter, createAsyncThunk } from '@reduxjs/toolkit';
import getDataRequest from '../utils/getDataRequest.js';
import getStringRequest from '../utils/getStringRequest.js';
import uploadBooks from '../utils/uploadBooks.js';

const booksAdapter = createEntityAdapter();

const initialState = booksAdapter.getInitialState({
  allBooks: [],
  totalItems: 0,
  query: '',
  typeSort: 'relevance',
  typeFilter: 'all',
  countItemsLoad: 0,
  selectedBook: null,
  loading: '',
  error: null,
});

const getBooks = (param) => {
  const stringRequest = getStringRequest(param);
  return getDataRequest(stringRequest);
};

export const searchBooks = createAsyncThunk('books/fetchData', getBooks);

const booksSlice = createSlice({
  name: 'books',
  initialState,
  reducers: {
    setTypeSort: (state, action) => {
      state.typeSort = action.payload;
    },
    setTypeFilter: (state, action) => {
      state.typeFilter = action.payload;
    },
    setQuery: (state, action) => {
      state.query = action.payload;
    },
    resetCount: (state) => {
      state.countItemsLoad = 0;
    },
    selectBook: (state, action) => {
      state.selectedBook = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(searchBooks.pending, (state) => {
        state.loading = 'loading';
        state.error = null;
      })
      .addCase(searchBooks.fulfilled, (state, action) => {
        const { totalItems, items } = action.payload;
        state.totalItems = totalItems;
        if (totalItems > 0) {
          state.allBooks = uploadBooks(items, state.allBooks, state.countItemsLoad);
          state.countItemsLoad = state.countItemsLoad + items.length;
        } else {
          state.allBooks = [];
          state.countItemsLoad = 0;
        }
        state.loading = 'success';
        state.error = null;
      })
      .addCase(searchBooks.rejected, (state, action) => {
        state.loading = 'failed';
        state.error = action.error;
        state.allBooks = [];
        state.countItemsLoad = 0;
        state.totalItems = 0;
      });
  },
});

export const {
  setTypeSort,
  setTypeFilter,
  setQuery,
  resetCount,
  selectBook,
} = booksSlice.actions;
export default booksSlice.reducer;
