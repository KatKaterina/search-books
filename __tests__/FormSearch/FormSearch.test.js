import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect.js';
import React, * as reactHooks from 'react';
import * as reduxHooks from 'react-redux';
import * as actions from '../../src/slices/booksSlice';
import FormSearch from '../../src/components/FormSearch';
import AppContext from '../../src/contexts/index.js';

jest.mock('react-redux');

const mockedUsedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUsedNavigate,
}));

const mockedDispatch = jest.spyOn(reduxHooks, 'useDispatch');
const mockedSelector = jest.spyOn(reduxHooks, 'useSelector');
const mockedNavigate = jest.spyOn(require('react-router-dom'), 'useNavigate');

const testValue = {
  query: '',
  typeFilter: 'all',
  typeSort: 'relevance',
  loading: '',
};
const step = 30;

describe('FormSearch', () => {
  it('create FormSearch', () => {
    mockedSelector.mockReturnValue(testValue);
    mockedDispatch.mockReturnValue(jest.fn());
    mockedNavigate.mockReturnValue(jest.fn());

    const component = render(
      <AppContext.Provider value={step}>
        <FormSearch />
      </AppContext.Provider>
    );

    expect(component).toMatchSnapshot();
  });

  /*it('dispatch actions', () => {
    // in progress
  });*/
});
