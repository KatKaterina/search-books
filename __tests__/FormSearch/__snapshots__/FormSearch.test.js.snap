// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormSearch create FormSearch 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="backdrop text-white container"
      >
        <div
          class="row"
        >
          <div
            class="mt-5 col"
          >
            <h1
              class="text-center"
            >
              Search for books
            </h1>
          </div>
        </div>
        <div
          class="row"
        >
          <form
            class=""
          >
            <div
              class="mb-1 input-group"
            >
              <input
                class="form-control"
                data-testid="search-field"
                placeholder="e.g. Little Prince"
                type="search"
                value=""
              />
              <button
                class="btn btn-light"
                data-testid="button-submit"
                type="submit"
              >
                üîç
              </button>
            </div>
            <div
              class="mb-5 mt-5 justify-content-center row"
            >
              <label
                class="form-label col-form-label col-md-auto"
              >
                Categories
              </label>
              <div
                class="col-md-auto"
              >
                <select
                  aria-label="select-categories"
                  class="form-select"
                  data-testid="select-categories"
                >
                  <option
                    data-testid="all"
                    value="all"
                  >
                    all
                  </option>
                  <option
                    data-testid="art"
                    value="art"
                  >
                    art
                  </option>
                  <option
                    data-testid="biography"
                    value="biography"
                  >
                    biography
                  </option>
                  <option
                    data-testid="computers"
                    value="computers"
                  >
                    computers
                  </option>
                  <option
                    data-testid="history"
                    value="history"
                  >
                    history
                  </option>
                  <option
                    data-testid="medicine"
                    value="medicine"
                  >
                    medicine
                  </option>
                  <option
                    data-testid="poetry"
                    value="poetry"
                  >
                    poetry
                  </option>
                </select>
              </div>
              <label
                class="form-label col-form-label col-md-auto"
              >
                Sorting by
              </label>
              <div
                class="col-md-auto"
              >
                <select
                  aria-label="select-sorting"
                  class="form-select"
                  data-testid="select-sorting"
                >
                  <option
                    data-testid="relevance"
                    value="relevance"
                  >
                    relevance
                  </option>
                  <option
                    data-testid="newest"
                    value="newest"
                  >
                    newest
                  </option>
                </select>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="backdrop text-white container"
    >
      <div
        class="row"
      >
        <div
          class="mt-5 col"
        >
          <h1
            class="text-center"
          >
            Search for books
          </h1>
        </div>
      </div>
      <div
        class="row"
      >
        <form
          class=""
        >
          <div
            class="mb-1 input-group"
          >
            <input
              class="form-control"
              data-testid="search-field"
              placeholder="e.g. Little Prince"
              type="search"
              value=""
            />
            <button
              class="btn btn-light"
              data-testid="button-submit"
              type="submit"
            >
              üîç
            </button>
          </div>
          <div
            class="mb-5 mt-5 justify-content-center row"
          >
            <label
              class="form-label col-form-label col-md-auto"
            >
              Categories
            </label>
            <div
              class="col-md-auto"
            >
              <select
                aria-label="select-categories"
                class="form-select"
                data-testid="select-categories"
              >
                <option
                  data-testid="all"
                  value="all"
                >
                  all
                </option>
                <option
                  data-testid="art"
                  value="art"
                >
                  art
                </option>
                <option
                  data-testid="biography"
                  value="biography"
                >
                  biography
                </option>
                <option
                  data-testid="computers"
                  value="computers"
                >
                  computers
                </option>
                <option
                  data-testid="history"
                  value="history"
                >
                  history
                </option>
                <option
                  data-testid="medicine"
                  value="medicine"
                >
                  medicine
                </option>
                <option
                  data-testid="poetry"
                  value="poetry"
                >
                  poetry
                </option>
              </select>
            </div>
            <label
              class="form-label col-form-label col-md-auto"
            >
              Sorting by
            </label>
            <div
              class="col-md-auto"
            >
              <select
                aria-label="select-sorting"
                class="form-select"
                data-testid="select-sorting"
              >
                <option
                  data-testid="relevance"
                  value="relevance"
                >
                  relevance
                </option>
                <option
                  data-testid="newest"
                  value="newest"
                >
                  newest
                </option>
              </select>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
